### PR TITLE
TSK-1824: Refactor createTask tests that involve Attachments; Throw exception in REST if taskId of an Attachment is incorrect

### PR DESCRIPTION
**Pull request by ryzheboka**
https://sonarcloud.io/summary/new_code?id=ryzheboka_taskana&branch=TSK-1824

---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    ObjectReference objRefValueNull =
+        taskService.newObjectReference("Company", "System", "Instance", "Type", null);
+    ObjectReference objRefCompanyNull =
+        taskService.newObjectReference(null, "System", "Instance", "Type", "Value");
+    List<Pair<String, ObjectReference>> valuesForTests =
+        List.of(
+            Pair.of("ObjRef is null", null),
+            Pair.of("Type of objRef is null", objRefTypeNull),
+            Pair.of("Value of objRef is null", objRefValueNull),
+            Pair.of("Company of objRef is null", objRefCompanyNull));
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:137]:
```suggestion
    List<Pair<String, ObjectReference>> valuesForTests =
      List.of(
          Pair.of("ObjRef is null", null),
          Pair.of("Type of objRef is null", defaultTestObjectReference().type(null).build()),
          Pair.of("Value of objRef is null", defaultTestObjectReference().value(null).build()),
          Pair.of(
              "Company of objRef is null", defaultTestObjectReference().company(null).build()));
```

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        p -> {
+          Task task = taskService.newTask(defaultWorkbasketSummary.getId());
+          task.setClassificationKey(defaultClassificationSummary.getKey());
+          task.setPrimaryObjRef(defaultObjectReference);
+          Attachment attachment = taskService.newAttachment();
+          attachment.setClassificationSummary(defaultClassificationSummary);
+          attachment.setObjectReference(p.getRight());
+          task.addAttachment(attachment);
+          assertThatThrownBy(() -> taskService.createTask(task))
+              .isInstanceOf(InvalidArgumentException.class);
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:149]:
```suggestion
          ObjectReference objectReference = p.getRight();
          Attachment attachment = taskService.newAttachment();
          attachment.setClassificationSummary(defaultClassificationSummary);
          attachment.setObjectReference(objectReference);
          Task task = taskService.newTask(defaultWorkbasketSummary.getId());
          task.setClassificationKey(defaultClassificationSummary.getKey());
          task.setPrimaryObjRef(defaultObjectReference);
          task.addAttachment(attachment);

          ThrowingCallable call = () -> taskService.createTask(task);

          assertThatThrownBy(call).isInstanceOf(InvalidArgumentException.class);
```

Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:149]: I know that the original tests did not verify the message of the IllegalArgumentException, but I am undecided whether we should verify the message of those Exceptions. What do you think?

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  @WithAccessId(user = "user-1-1")
+  @TestFactory
+  Stream<DynamicTest>
+      should_ThrowException_When_CreatingTaskWithInvalidClassificationOfAttachment() {
+    Classification nonExistingClassification =
+        classificationService.newClassification(
+            "123key345", defaultWorkbasketSummary.getDomain(), "non existing type");
+    Classification classificationWithoutKey =
+        classificationService.newClassification(
+            null, defaultClassificationSummary.getDomain(), defaultClassificationSummary.getType());
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:164]: these have to be ClassificationSummaries

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    return DynamicTest.stream(valuesForTests.iterator(), Pair::getLeft, test);
+  }
+
+  @WithAccessId(user = "user-1-1")
+  @TestFactory
+  Stream<DynamicTest>
+      should_ThrowException_When_CreatingTaskWithInvalidClassificationOfAttachment() {
+    Classification nonExistingClassification =
+        classificationService.newClassification(
+            "123key345", defaultWorkbasketSummary.getDomain(), "non existing type");
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:161]: I prefer a more obvious key like `NOT_EXISTING`

Comment (mustaphazorgati) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:161]: you can hardcode `DOMAIN_A` since they exist in the configuration

Comment (mustaphazorgati) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:161]: please use a proper type to clarify that we're only testing the key here.

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 3 above)
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:164]: please hardcode domain and type :)

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+          Task task = taskService.newTask(defaultWorkbasketSummary.getId());
+          task.setClassificationKey(defaultClassificationSummary.getKey());
+          task.setPrimaryObjRef(defaultObjectReference);
+          Attachment attachment = taskService.newAttachment();
+          attachment.setClassificationSummary(q.getSecond());
+          attachment.setObjectReference(defaultObjectReference);
+          task.addAttachment(attachment);
+          assertThatThrownBy(() -> taskService.createTask(task))
+              .isInstanceOf(q.getFourth())
+              .hasMessageContaining(q.getThird());
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:196]: separate using the given-when-then theorem

Comment (mustaphazorgati) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:196]: rearrange, so that Attachment gets created before Task (so that all setters of the Task are in close proximity)

Comment (mustaphazorgati) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:196]: create variables for each value of the quadruple.

Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:196]: We have an issue: The `ClassificationNotFoundException` contains some fields which have to be verified while the `InvalidArgumentException` contains different messages. 
You've parameterized the Exception class and the message. Why don't you just parameterize an instance of the exception and use the recursivecomparison feature from assertj?

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                defaultClassificationSummary.getKey(),
+                defaultClassificationSummary.getDomain(),
+                defaultClassificationSummary.getType())
+            .asSummary();
+    attachment.setClassificationSummary(classification);
+    Task taskToCreate = taskService.newTask(defaultWorkbasketSummary.getId());
+    taskToCreate.setClassificationKey(defaultClassificationSummary.getKey());
+    taskToCreate.setPrimaryObjRef(defaultObjectReference);
+    taskToCreate.addAttachment(attachment);
+    taskToCreate = taskService.createTask(taskToCreate);
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:116]: please separate given and when :wink:

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    Attachment copiedAttachment =
+        taskService.getTask(defaultTaskWithAttachment.getId()).getAttachments().get(0).copy();
+    Task taskToCreate = taskService.newTask(defaultWorkbasketSummary.getId());
+    taskToCreate.setClassificationKey(defaultClassificationSummary.getKey());
+    taskToCreate.setPrimaryObjRef(defaultObjectReference);
+    taskToCreate.addAttachment(copiedAttachment);
+    Task result = taskService.createTask(taskToCreate);
+
+    assertThat(result.getAttachments().get(0).getTaskId()).isEqualTo(result.getId());
+  }
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:97]: This does not test TSK-1824 since the `copiedAttachment` does not contain a different taskId.
Please don't copy the Attachment. Just reuse the Attachment from the `defaultTaskWithAttachment`

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            .classificationSummary(defaultClassificationSummary)
+            .attachments(defaultAttachment)
+            .buildAndStore(taskService, "user-1-1");
+  }
+
+  @WithAccessId(user = "user-1-1")
+  @Test
+  void should_SetTaskIdOfAttachmentCorrectly_WhenCopyingAttachment() throws Exception {
+    Attachment copiedAttachment =
+        taskService.getTask(defaultTaskWithAttachment.getId()).getAttachments().get(0).copy();
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:89]:
```suggestion
        defaultTaskWithAttachment.getAttachments().get(0).copy();
```

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            .classificationSummary(defaultClassificationSummary)
+            .objectReference(defaultObjectReference)
+            .build();
+    defaultTaskWithAttachment =
+        TaskBuilder.newTask()
+            .primaryObjRef(defaultObjectReference)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .classificationSummary(defaultClassificationSummary)
+            .attachments(defaultAttachment)
+            .buildAndStore(taskService, "user-1-1");
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:82]: please be consistent and static import all test-api classes :)

btw: should we make that a guideline?

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 9 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:82]: I will include this in the guidelines :)

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 10]
```diff
@@ -416,11 +418,23 @@ public ResponseEntity<TaskRepresentationModel> updateTask(
...
   }
 
+  private void verifyTaskIdOfAttachments(
+      List<AttachmentRepresentationModel> attachments, String taskId)
+      throws InvalidArgumentException {
+    for (AttachmentRepresentationModel attachment : attachments) {
+      if (attachment.getTaskId() != null && !attachment.getTaskId().equals(taskId)) {
+        throw new InvalidArgumentException("TaskId of Attachment should be empty.");
+      }
+    }
```
[the remark this answers is not in the document]
Comment (ryzheboka) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:435]: I would like to leave it as is. Throwing exceptions inside a stream could be troublesome

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 9 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:82]: What do you think about static imports of DefaultTestEntities instead of all test-api classes?

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 8 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:89]: Ok :)

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 7 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:97]: Ok :)

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 1 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:137]: Ok :)

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 2 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:149]: Ok :)

Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:149]: I will verify the message

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 3 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:164]: Ok :)

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 11]
```diff
@@ -677,6 +678,29 @@ void should_NotGetEmptyObjectReferencesList_When_GettingTaskWithObjectReferences
     assertThat(repModel.getSecondaryObjectReferences()).isNotEmpty();
   }
 
+  @Test
+  void should_ThrowException_When_AttachmentAlreadyHasTaskId() {
```
[the remark this answers is not in the document]
Comment (ryzheboka) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:682]: I will add it :)

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:435]: You wouldn't throw the exception inside the Stream API
```
if (attachments.stream()
        .map(AttachmentRepresentationModel::getTaskId)
        .filter(Objects::nonNull)
        .anyMatch(not(taskId::equals))) {
      throw new InvalidArgumentException("TaskId of Attachment should be empty.");
    }
```

[the remark this answers is not in the document]
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:435]: Or any variation of this :D

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 4 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:156]: Ok :)

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 3 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:159]: Ok :)

File: lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java (hunk 5 above)
Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:191]: I will create the variables

Comment (ryzheboka) [reply lib/taskana-core-test/src/test/java/acceptance/task/create/CreateTaskWithAttachmentAccTest.java:191]: Ok :)

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ryzheboka) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:435]: It looks great, I will implement it this way :)

**Comment by ryzheboka** [comment]
Closed by #2355
